include/dkm_matrix.hpp

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#ifndef DKM_MATRIX_HPP
+#define DKM_MATRIX_HPP
+
+#include <vector>
+#include <cassert>
+#include <algorithm>
+
+namespace dkm
+{
+// This class is only an interface! Not designed to be used outside library internals.
+
+template<typename T>
+class as_matrix
+{
+    const T *data = nullptr;
+
+    // column major indexer
+    auto cm_indexer(size_t i, size_t j) const -> const T&
+    {
+        assert(i >= 0 && j >= 0 && i < n_rows && j < n_cols);
+        return data[j * n_rows + i];
+    }
+
+     // row major indexer
+    auto rm_indexer(size_t i, size_t j) const -> const T&
+    {
+        assert(i >= 0 && j >= 0 && i < n_rows && j < n_cols);
+        return data[i * n_cols + j];
+    }
+
+    const T& (as_matrix<T>::*indexer) (size_t, size_t) const = nullptr;
+
+public:
+    const size_t n_rows, n_cols;
+
+    as_matrix(const T *data, size_t n_rows, size_t n_cols, bool col_major = true)
```

Comment:
It would be great to see a constructor for conversion from the existing `vector<array<T, N>>` type as well to ease migration for existing users.

Reply:
We can also provide an overload and mark it as `[[deprecated]]`. What do you think?

Reply:
This approach is also acceptable, as long as [[deprecated]] will be ignored by older compilers.